ext/nnstreamer/tensor_filter/tensor_filter_edgetpu.cc

```diff
@@ -0,0 +1,395 @@
+ * @see		http://github.com/nnsuite/nnstreamer
+ * @author	MyungJoo Ham <myungjoo.ham@samsung.com>
+ * @bug		No known bugs except for NYI items
+ *
+ * This is the per-NN-framework plugin (Edge TPU) for tensor_filter.
+ *
+ * @todo This supports single-model-single-TPU only. Prep for multi-TPU.
+ * @todo A lot of this duplicate tf-lite filter.
+ *       We may be able to embed this code into tf-lite filter code.
+ */
+#include <iostream>
+
+#include <stdint.h>
+
+#include <nnstreamer_plugin_api_filter.h>
+#include <tensor_common.h>
+#include <glib.h>
+
+#include <edgetpu.h>
+#include <tensorflow/lite/model.h>
+#include <tensorflow/lite/interpreter.h>
+#include <tensorflow/lite/builtin_op_data.h>
+#include <tensorflow/lite/kernels/register.h>
+
+void init_filter_edgetpu (void) __attribute__ ((constructor));
+void fini_filter_edgetpu (void) __attribute__ ((destructor));
+
+/**
+ * @brief Internal data structure
+ */
+typedef struct {
+  char *model_path; /**< The model *.tflite file */
+  GstTensorsInfo input; /**< Input tensors metadata */
+  GstTensorsInfo output; /**< Output tensors metadata */
+
+  /* EdgeTPU + Tensorflow-lite Execution */
+  std::unique_ptr<tflite::Interpreter> model_interpreter;
+      /**< TFLite interpreter */
+  tflite::Interpreter *interpreter;
+      /**< model_interpreter.get() */
+  std::shared_ptr<edgetpu::EdgeTpuContext> edgetpu_context;
+      /**< EdgeTPU Device context */
+  std::unique_ptr<tflite::FlatBufferModel> model;
+      /**< Loaded TF Lite model (from model_path) */
+} pdata;
+
+/**
+ * @brief Get TF-Lite interpreter w/ edgetpu context
+ */
+static std::unique_ptr<tflite::Interpreter>
+BuildEdgeTpuInterpreter(const tflite::FlatBufferModel &model,
+      edgetpu::EdgeTpuContext* edgetpu_context)
+{
+  tflite::ops::builtin::BuiltinOpResolver resolver;
+  resolver.AddCustom(edgetpu::kCustomOp, edgetpu::RegisterCustomOp());
+  std::unique_ptr<tflite::Interpreter> interpreter;
+  if (tflite::InterpreterBuilder(model, resolver)(&interpreter) !=
+      kTfLiteOk) {
+    std::cerr << "Failed to build interpreter." << std::endl;
+  }
+
+  interpreter->SetExternalContext(kTfLiteEdgeTpuContext, edgetpu_context);
+  interpreter->SetNumThreads(1);
+  if (interpreter->AllocateTensors() != kTfLiteOk) {
+    std::cerr << "Failed to allocate tensors." << std::endl;
+  }
+
+  return interpreter;
+}
+
+static void
+edgetpu_close (const GstTensorFilterProperties * prop, void **private_data);
+
+/**
+ * @brief Configure private_data
+ */
+static int
+allocateData (const GstTensorFilterProperties * prop, void **private_data)
+{
+  pdata *data;
+  if (*private_data != NULL) {
+    /* Already opened */
+    data = (pdata *) *private_data;
+    if (data->model_path && g_strcmp0 (prop->model_files[0],
+            data->model_path)) {
+      edgetpu_close (prop, private_data); /* Opening different file! */
+    } else {
+      return 1; /* Already opened with same file. Skip ops */
```

Comment:
I think the condition should be like this:
```
if (data->model_path && g_strcmp0 (prop->model_files[0], data->model_path) == 0) {
   return 1;
} else {
   edgetpu_close (prop, private_data);
}
```

because, if `data->model_path` is NULL, this function returns always 1

Reply:
Fixed.